### PR TITLE
Issue #516: Separate REST API web front

### PR DESCRIPTION
**Pull request by tim-bezhashvyly**

### CODE REVIEW

File: src/RestApi/ApiRequestHandlerLocator.php [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 namespace LizardsAndPumpkins\RestApi;
 
+use LizardsAndPumpkins\Http\Routing\HttpRequestHandler;
+
 class ApiRequestHandlerLocator
```
Comment (tim-bezhashvyly) [inline src/RestApi/ApiRequestHandlerLocator.php]: I still don't like the fact of having 2 chains. 😞

File: src/RestApiWebFront.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            $statusCode = $response->getStatusCode();
+        } catch (\Exception $e) {
+            $body = json_encode(['error' => $e->getMessage()]);
+            $statusCode = HttpResponse::STATUS_BAD_REQUEST;
+        }
+
+        return $this->createJsonResponse($body, $statusCode);
+    }
+
+    protected function createMasterFactory(): MasterFactory
```
Comment (tim-bezhashvyly) [inline src/RestApiWebFront.php]: Not final for testing purposes.

File: src/RestApiWebFront.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        $masterFactory->register(new ProductSearchFactory());
+        $masterFactory->register(new UpdatingProductImportCommandFactory());
+        $masterFactory->register(new UpdatingProductImageImportCommandFactory());
+        $masterFactory->register(new UpdatingProductListingImportCommandFactory());
+        $masterFactory->register($this->getImplementationSpecificFactory());
+    }
+
+    final protected function registerRouters(HttpRouterChain $routerChainChain)
+    {
+        $routerChainChain->register($this->getMasterFactory()->createApiRouter());
```
Comment (tim-bezhashvyly) [inline src/RestApiWebFront.php]: Chain of a single element. 😞

File: src/RestApiWebFront.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace LizardsAndPumpkins;
```
Comment (tim-bezhashvyly) [inline src/RestApiWebFront.php]: I've put it to the root namespace same as `DefaultWebFront` but maybe `\LizardsAndPumpkins\RestApi` would be a better place.

File: tests/Integration/Suites/InjectableRestApiWebFront.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+declare(strict_types=1);
+
+namespace LizardsAndPumpkins;
+
+use LizardsAndPumpkins\Http\HttpRequest;
+use LizardsAndPumpkins\Util\Factory\Factory;
+use LizardsAndPumpkins\Util\Factory\MasterFactory;
+
+class InjectableRestApiWebFront extends RestApiWebFront
```
Comment (tim-bezhashvyly) [inline tests/Integration/Suites/InjectableRestApiWebFront.php]: The only difference from `InjectableDefaultWebFront` is the extended class. 😞

File: tests/Integration/Suites/ProductSearchEdgeToEdgeTest.php [hunk 6]
```diff
@@ -55,6 +57,11 @@ function () {
         );
     }
 
+    final protected function setUp()
+    {
+        $this->importProductListingTemplateFixtureViaApi();
```
Comment (tim-bezhashvyly) [inline tests/Integration/Suites/ProductSearchEdgeToEdgeTest.php]: Wondering how it was working at all.

File: src/RestApiWebFront.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        $masterFactory->register($this->getImplementationSpecificFactory());
+    }
+
+    final protected function registerRouters(HttpRouterChain $routerChainChain)
+    {
+        $routerChainChain->register($this->getMasterFactory()->createApiRouter());
+    }
+
+    private function createJsonResponse(string $body, int $statusCode): HttpResponse
+    {
```
Comment (tim-bezhashvyly) [inline src/RestApiWebFront.php]: My thought was that REST API response can not contain any headers but no problem, I will merge them.

Comment (tim-bezhashvyly) [inline src/RestApiWebFront.php]: Done.

File: tests/Integration/Suites/InjectableRestApiWebFront.php (hunk 5 above)
Comment (tim-bezhashvyly) [inline tests/Integration/Suites/InjectableRestApiWebFront.php]: I would leave it as it is.

**Comment by tim-bezhashvyly** [comment]
Ready for next review.

File: src/RestApi/RestApiFactory.php [hunk 8]
```diff
@@ -31,7 +31,7 @@ public function createApiRouter(): ApiRouter
         return new ApiRouter($this->getApiRequestHandlerLocator());
     }
 
-    public function createCatalogImportApiV1PutRequestHandler(): CatalogImportApiV1PutRequestHandler
+    public function createCatalogImportApiV1PutRequestHandler(): CatalogImportApiV2PutRequestHandler
```
Comment (Vinai) [inline src/RestApi/RestApiFactory.php]: Why this change? If it was based on a change request from me I'm sorry I lost the thread of thought.
Right now I find it confusing that the method name and the return type don't match.

File: src/RestApiWebFront.php (hunk 7 above)
Comment (Vinai) [inline src/RestApiWebFront.php]: Thanks!

File: src/RestApi/RestApiFactory.php (hunk 8 above)
Comment (tim-bezhashvyly) [inline src/RestApi/RestApiFactory.php]: In this case PHPStorm marks it all yellow as `CatalogImportApiV1PutRequestHandler` delegates `create` constructor to `CatalogImportApiV2PutRequestHandler`.

Comment (Vinai) [inline src/RestApi/RestApiFactory.php]: I don't like changing code in a "wrong" way because if flaws in the PHPStorm PHP analyzer.

Comment (Vinai) [inline src/RestApi/RestApiFactory.php]: Oh well, lets leave it.

**Comment by Vinai** [comment]
Can you rebase the PR please?
